### PR TITLE
EN-2314 EN-2289 DE instances authentication, configuration and connection test

### PR DESCRIPTION
**Pull request by zonia3000**

### CODE REVIEW

**Comment by entando-service** [comment]
All is well

**Comment by zonia3000** [comment]
For performing the authentication I replaced the `RestTemplate` with `OAuth2RestTemplate`. This class automatically retrieves an OAuth2 token when it is missing or expired and then performs the desired request. Each `OAuth2RestTemplate` stores the token inside of it, so it is necessary to have a different instance of this class for each DE instance. This means also that these objects need to be added or replaced every time a DE configuration changes. This complicated a bit the `DigitalExchangesServiceImpl` so I preferred to create also the manager layer and move this logic there, keeping on the service only the validation.
There is however a positive consequence of this: having a different `RestTemplate` for each DE instance eases the configuration of a different timeout for each DE instance, so the DE client is now simpler.

File: engine/src/test/java/org/entando/entando/web/digitalexchange/DigitalExchangesControllerIntegrationTest.java [hunk 1]
```diff
@@ -148,9 +171,21 @@ public void shouldFailDeletingDigitalExchangeBecauseNotFound() throws Exception
...
 
+    @Test
+    public void shouldTestInstance() throws Exception {
+
+        ResultActions result = createAuthRequest(get(BASE_URL + "/test/{name}", "DE 1")).execute();
+
+        result.andExpect(status().isOk());
+        result.andExpect(jsonPath("$.metaData").isEmpty());
+        result.andExpect(jsonPath("$.errors").isEmpty());
+        result.andExpect(jsonPath("$.payload", is("OK")));
```
Comment (rzsouza) [inline engine/src/test/java/org/entando/entando/web/digitalexchange/DigitalExchangesControllerIntegrationTest.java]: Nothing wrong here, but if you're feeling lazy next time:

```
        result.andExpect(status().isOk())
                .andExpect(jsonPath("$.metaData").isEmpty())
                .andExpect(jsonPath("$.errors").isEmpty())
                .andExpect(jsonPath("$.payload", is("OK")));

```

File: engine/src/main/java/org/entando/entando/aps/system/services/digitalexchange/DigitalExchangesManagerImpl.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+ * details.
+ */
+package org.entando.entando.aps.system.services.digitalexchange;
+
+import com.agiletec.aps.system.SystemConstants;
+import com.agiletec.aps.system.exception.ApsSystemException;
+import com.agiletec.aps.system.services.baseconfig.ConfigInterface;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.util.ArrayList;
```
[the remark this answers is not in the document]
Comment (zonia3000) [reply engine/src/main/java/org/entando/entando/aps/system/services/digitalexchange/DigitalExchangesManagerImpl.java]: fixed

File: engine/src/test/java/org/entando/entando/aps/system/services/digitalexchange/DigitalExchangesServiceTest.java [hunk 3]
```diff
@@ -122,28 +99,57 @@ public void shouldFailUpdatingDigitalExchange() {
...
-        when(configManager.getConfigItem(SystemConstants.CONFIG_ITEM_DIGITAL_EXCHANGES))
-                .thenReturn("<digitalExchanges>"
-                        + "  <digitalExchange>"
-                        + "    <name>DE 1</name>"
-                        + "    <url>http://www.entando.com/</url>"
-                        + "  </digitalExchange>"
-                        + "</digitalExchanges>");
+        when(manager.findByName(any(String.class)))
+                .thenAnswer(invocation -> {
+                    String name = (String) invocation.getArgument(0);
```
[the remark this answers is not in the document]
Comment (zonia3000) [reply engine/src/test/java/org/entando/entando/aps/system/services/digitalexchange/DigitalExchangesServiceTest.java]: fixed

File: engine/src/test/java/org/entando/entando/web/digitalexchange/DigitalExchangesControllerIntegrationTest.java (hunk 1 above)
Comment (zonia3000) [reply engine/src/test/java/org/entando/entando/web/digitalexchange/DigitalExchangesControllerIntegrationTest.java]: thanks for the suggestion

File: engine/src/test/java/org/entando/entando/aps/system/services/digitalexchange/client/DigitalExchangesClientTest.java [hunk 4]
```diff
@@ -53,10 +55,7 @@
...
 
     @Mock
-    private RestTemplate restTemplate;
-
-    @Mock
-    private DigitalExchangesService service;
+    private DigitalExchangesManager manager;
 
     @Mock
     private MessageSource messageSource;
```
[the remark this answers is not in the document]
Comment (zonia3000) [reply engine/src/test/java/org/entando/entando/aps/system/services/digitalexchange/client/DigitalExchangesClientTest.java]: In this class it is used, but the same mock was unused in the service class, so I suppose you picked the wrong file in the review.

**Comment by entando-service** [comment]
All is well

**Pull request merged**